nondex-common/src/main/java/edu/illinois/nondex/common/SystematicRandom.java

```diff
@@ -0,0 +1,142 @@
+"Software"), to deal in the Software without restriction, including
+without limitation the rights to use, copy, modify, merge, publish,
+distribute, sublicense, and/or sell copies of the Software, and to
+permit persons to whom the Software is furnished to do so, subject to
+the following conditions:
+
+The above copyright notice and this permission notice shall be
+included in all copies or substantial portions of the Software.
+
+THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
+EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
+MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
+NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE
+LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION
+OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION
+WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
+*/
+
+package edu.illinois.nondex.common;
+
+import java.io.BufferedWriter;
+import java.io.File;
+import java.io.IOException;
+import java.nio.charset.Charset;
+import java.nio.charset.StandardCharsets;
+import java.nio.file.Files;
+import java.nio.file.Paths;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Random;
+import java.util.Stack;
+
+public class SystematicRandom extends Random {
+    public static int STARTING_COUNT = 59;
+    int replayIndex;
+    int count;
+    Stack<StackElement> choice;
+    String logFileName;
+    Charset utf8 = StandardCharsets.UTF_8;
+    List<String> lines;
+
+    public SystematicRandom() {
+        logFileName = System.getenv("logFileName");
+        if (logFileName != null) {
+            File file = new File(this.logFileName);
+            if (!file.exists()) {
+                choice = new Stack<StackElement>();
+                replayIndex = 0;
+                count = 0;
+            } else {
+                choice = new Stack<StackElement>();
+                try {
+                    lines = Files.readAllLines(Paths.get(logFileName));
+                } catch (IOException ioe) {
+                    ioe.printStackTrace();
+                }
+                Object[] choiceValues = new Object[3];
+                for (String element: lines) {
+                    String delimit = "[ ]+";
+                    if (!element.isEmpty()) {
+                        choiceValues = element.split(delimit);
+                        int last = Integer.parseInt(choiceValues[0].toString());
+                        int max = Integer.parseInt(choiceValues[1].toString());
+                        boolean explore = Boolean.parseBoolean(choiceValues[2].toString());
+                        StackElement lastMax = new StackElement( last, max, explore );
+                        choice.push(lastMax);
+                    }
+                }
+            }
+        } else {
+            System.err.println("The env variable does not exist. File could not be created");
+            System.exit(1);
```

Comment:
This appears to be the problem that's causing the build to fail on Travis (and any machine where the environment variable is not defined). The System.exit(1) is invoked since the environment variable does not exist, and due to how the tests are doing things with threads and stuff, it just hangs, leading to the timeout on Travis. There really should be no System.exit(1) in the code, some exception should be thrown and propagated upwards.